Bluetooth: Host: Run tx_processor on its own thread

alwa-nordic opened the pull request:
Move tx_processor off the system work queue to its own thread. Solves deadlock where `bt_hci_cmd_alloc()` blocks the system work queue.

alwa-nordic committed:
Bluetooth: Host: Don't call user callback from TX thread

ATT is invoking user callbacks in its net_buf destroy function. It is
common practice that these callbacks can block on bt_hci_cmd_alloc().
This is a deadlock when the net_buf_unref() happens inside the HCI
driver, invoked from tx_processor.

Blocking callbacks like this appear in our own samples. See further down
about how this problem was detected.

tx_processor not protect against blocking callbacks so it is de-facto
forbidden. The Host should not equip net_bufs with dangerous destroy
callbacks.

This commit makes ATT defer its net_buf destruction and user callback
invocation to the system workqueue, so that net_buf_unref is safe to
call from non-blocking threads. In the case of the deadlock, the
net_buf_unref() was below the tx_processor in the call stack, which (at
the time of this commit) is on the system work queue, so defering it to
the system work queue is preserving the existing behavior.

Future improvement may be to allow the user to provide their own
workqueue for ATT callbacks.

This deadlock was detected because the following test was failing while
moving tx_processor to the bt_taskq:

    tests/bsim/bluetooth/ll/throughput/tests_scripts/gatt_write.sh

The above test has an ATT callback `write_cmd_cb` invokes
`bt_conn_le_param_update` can block waiting for `tx_processor`.

The reason it was not failing while tx_processor was on the system work
queue is that the GATT API has a special non-blocking behavior when
called from the system work queue.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic committed:
Bluetooth: Samples: Reduce RAM requirement of peripheral_identity

Reduce BT_MAX_CONN from 62 to 61 to make it build on integration
platform qemu_cortex_m3/ti_lm3s6965 when we add bt_taskq in subsequent
commit.

The number 62 seems arbitrary here, so reducing it by one should not
have any practical impact.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

cvinayak commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -132,6 +132,28 @@ config BT_DRIVER_RX_HIGH_PRIO
+	# Hidden unconfigurable option
+	bool
+	default y if !SOC_SERIES_NRF51X
+	# This option is automatically selected for all platforms except nRF51
```
The Zephyr Controller's redesigned control procedure max. CPU usage due to handling of tx ack and the processing of the Rx-ed PDU contiguously is higher within the 150 us tIFS. Hence, under controlled conditions nRF51 assertion checks fail detecting the increased CPU usage that is causing violation of hard real-time deadlines.

@alwa-nordic what is the reason for `default y if !SOC_SERIES_NRF51X` ? If this is assertions in the Controller, then you do not need this (and can be documented as Controller's known issue). If this is due to RAM or Flash usage, then on the SoC variant need not be using in `samples.yaml` that fail build.

alwa-nordic replied:
@jhedberg, would you like to move this discussion to an issue or other forum?

PavelVPV reviewed (commented):
Sorry, missed that we need to disable this code if tx processor thread is enabled:
https://github.com/zephyrproject-rtos/zephyr/blob/27904e9b135aeb13e132c4d117c72f5caa4eebb5/subsys/bluetooth/host/hci_core.c#L473-L500

PavelVPV commented:
> > Sorry, missed that we need to disable this code if tx processor thread is enabled:
> > https://github.com/zephyrproject-rtos/zephyr/blob/27904e9b135aeb13e132c4d117c72f5caa4eebb5/subsys/bluetooth/host/hci_core.c#L473-L500
> 
> Should we disable it? Or keep it and print a warning like in [93c2d8d](https://github.com/zephyrproject-rtos/zephyr/commit/93c2d8ddbab8925fc73948364976156e463ab002)? It may be a good idea to keep it until we do a full audit of the code on tx_processor and determine that it cannot possibly happen.

I think this path should never exist in case of tx processor thread. If it happens that somehow Host calls `bt_hci_cmd_send_sync` from tx processor, `bt_hci_cmd_send_sync` will deadlock as semaphore is released after a command is sent and response is received. But since the command is never sent, the response will never be received. A warning may not be noticed, but deadlock will be reported and gets fixed.

alwa-nordic committed:
Bluetooth: Host: Disable bt_hci_cmd_send_sync workaround

The workaround in bt_cmd_send_sync should no longer by needed when
tx_processor is not on the system work queue.

Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

cvinayak commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -132,6 +132,28 @@ config BT_DRIVER_RX_HIGH_PRIO
+
+config BT_TX_PROCESSOR_THREAD_PRIO
+	int
+	default -1
```
> detect transaction violations in ATT and SMP protocols.

This is to prevent remote initiated starvation of local ACL Tx buffers, i.e. example, a remote peer sending a burst of ATT Read Requests violating transaction rules leading to local enqueuing back ATT Read/Error Responses.

Hence, even if do not have transaction violation detection today, to be consistent with `ONFIG_BT_HCI_TX_PRIO` we should use `-7`. Another PR or GH issue can address collective the use of co-operative priority values like 6, 7 and 8 currently in the Bluetooth subsystem.

PavelVPV replied:
TX processor has been running on sysworkq before which priority is -1. Can we do profiling and address the priority issue in a separate PR? It is hard to change something without having evidence that this change is needed (even if it is needed).

alwa-nordic replied:
> This is to prevent remote initiated starvation of local ACL Tx buffers, i.e. example, a remote peer sending a burst of ATT Read Requests violating transaction rules leading to local enqueuing back ATT Read/Error Responses.

Well, we don't have those checks in `att.c` now. So changing the priority of tx_processor is pointless.

alwa-nordic replied:
OK, if I understand correctly, the final compromise is to inherit the system work queue priority and build assert if that's in the preemptive range. (I'm getting some design-by-committee vibes here, but OK :sweat_smile:)

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -132,6 +132,27 @@ config BT_DRIVER_RX_HIGH_PRIO
+	bool
+	# This option is automatically selected for all platforms except nRF51
+	# due to limited RAM on nRF51 devices.
+	default n if !SOC_SERIES_NRF51X
```
Isn't the negation incorrect here? I.e. you *don't* want this on nRF51? Maybe it'd be simpler to just have a single `default y if !SOC_SERIES_NRF51X` line?

wearyzen commented on samples/bluetooth/peripheral_identity/prj.conf:
```diff
@@ -1,12 +1,13 @@
-CONFIG_BT_MAX_CONN=62
-CONFIG_BT_ID_MAX=62
+CONFIG_BT_MAX_CONN=61
+CONFIG_BT_ID_MAX=61
```
I think board/soc vendors should be the maintainer, TI in this case but I see [here](https://github.com/zephyrproject-rtos/zephyr/blob/a8bf08bf4e2f43ab915ae35b0a4e7418f7f978a8/MAINTAINERS.yml#L4884C7-L4884C13) that the soc is also not maintained. I would reach out to someone in TI who might be aware of this soc.

alwa-nordic replied:
@cvinayak, did you get the answer to your question? I'm unable to contact the maintainer, as there is none. Is this thread a NAK from you? If so, can you help out moving it forwards?

Thalley commented on subsys/bluetooth/host/att.c:
Coding guidelines require that functions that may return an error, shall have the return value checked (`k_work_submit_to_queue` may return an error)

Thalley commented on subsys/bluetooth/host/hci_core.c:
Coding guidelines require that functions that may return an error, shall have the return value checked (`k_work_submit_to_queue` may return an error)

alwa-nordic replied:
OK, I'll assert and document.

Thalley reviewed (commented):
Overall looks good. A few trivial formatting comments.

alwa-nordic committed:
Bluetooth: Host: Run tx processor on its own thread

When thread that TX processor is used to send commands and data to
Controller is also used for sync commands sending and command buffer
allocation, a deadlock happens.

This thread is used to avoid such deadlocks by moving TX processor to
its own dedicated thread exclusively used by tx processor only.

Co-authored-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
Signed-off-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

alwa-nordic commented:
Pushed the typo fix. This PR ready now.

alwa-nordic commented on subsys/bluetooth/host/hci_core.c:
I mean to refer to any and all errors from `k_work_submit_to_queue`. I detailed above why none of the documented errors apply, which justifies the assertion. But here I also justify that nothing needs to be cleaned up if the assertion fails. There is a possibility someone else will repeat the call, which will resolve the error condition, so I'm thinking for defensive coding it's better to ignore the error rather than `k_oops()`. What do you think?
